src/operations/client_bulk_write/command_builder.ts

```diff
@@ -0,0 +1,233 @@
+import { type Document } from '../../bson';
+import { DocumentSequence } from '../../cmap/commands';
+import type {
+  AnyClientBulkWriteModel,
+  ClientBulkWriteOptions,
+  ClientDeleteManyModel,
+  ClientDeleteOneModel,
+  ClientInsertOneModel,
+  ClientReplaceOneModel,
+  ClientUpdateManyModel,
+  ClientUpdateOneModel
+} from './common';
+
+/** @internal */
+export interface ClientBulkWriteCommand {
+  bulkWrite: 1;
+  errorsOnly: boolean;
+  ordered: boolean;
+  ops: DocumentSequence;
+  nsInfo: DocumentSequence;
+  bypassDocumentValidation?: boolean;
+  let?: Document;
+}
+
+/** @internal */
+export class ClientBulkWriteCommandBuilder {
+  models: AnyClientBulkWriteModel[];
+  options: ClientBulkWriteOptions;
+
+  /**
+   * Create the command builder.
+   * @param models - The client write models.
+   */
+  constructor(models: AnyClientBulkWriteModel[], options: ClientBulkWriteOptions) {
+    this.models = models;
+    this.options = options;
+  }
+
+  /**
+   * Gets the errorsOnly value for the command, which is the inverse of the
+   * user provided verboseResults option. Defaults to true.
+   */
+  get errorsOnly(): boolean {
+    if ('verboseResults' in this.options) {
+      return !this.options.verboseResults;
+    }
+    return true;
+  }
+
+  /**
+   * Build the bulk write commands from the models.
+   */
+  buildCommands(): ClientBulkWriteCommand[] {
+    // Iterate the models to build the ops and nsInfo fields.
+    const operations = [];
+    let currentNamespaceIndex = 0;
+    const namespaces = new Map<string, number>();
+    for (const model of this.models) {
+      const ns = model.namespace;
+      if (namespaces.has(ns)) {
+        operations.push(buildOperation(model, namespaces.get(ns) as number));
+      } else {
+        namespaces.set(ns, currentNamespaceIndex);
+        operations.push(buildOperation(model, currentNamespaceIndex));
+        currentNamespaceIndex++;
+      }
+    }
+
+    const nsInfo = Array.from(namespaces.keys()).map(ns => {
+      return { ns: ns };
+    });
+
+    // The base command.
+    const command: ClientBulkWriteCommand = {
+      bulkWrite: 1,
+      errorsOnly: this.errorsOnly,
+      ordered: this.options.ordered ?? true,
+      ops: new DocumentSequence(operations),
+      nsInfo: new DocumentSequence(nsInfo)
+    };
+    // Add bypassDocumentValidation if it was present in the options.
+    if ('bypassDocumentValidation' in this.options) {
+      command.bypassDocumentValidation = this.options.bypassDocumentValidation;
+    }
+    // Add let if it was present in the options.
+    if ('let' in this.options) {
+      command.let = this.options.let;
+    }
```

Comment:
https://github.com/mongodb/node-mongodb-native/blob/NODE-6327/src/operations/aggregate.ts#L121-L127

It's a little strange to me that we've switched up the check to ingest these arguments. The code I link to isn't perfect either by only using truthiness, but this check seems to go out of its way to put `let: null` onto the command if that is present on the options. Is that intentional?

Reply:
I updated the check for `let` to not allow null as in the example linked. Nice catch.